core/tracing/hooks.go

```diff
@@ -68,7 +71,26 @@ type BlockEvent struct {
 	Safe      *types.Header
 }
 
+// Backend interface provides the common API services (that are provided by
+// both full and light clients) with access to necessary functions.
+type Backend interface {
+	HeaderByHash(ctx context.Context, hash common.Hash) (*types.Header, error)
+	HeaderByNumber(ctx context.Context, number rpc.BlockNumber) (*types.Header, error)
+	BlockByHash(ctx context.Context, hash common.Hash) (*types.Block, error)
+	BlockByNumber(ctx context.Context, number rpc.BlockNumber) (*types.Block, error)
+	GetTransaction(ctx context.Context, txHash common.Hash) (bool, *types.Transaction, common.Hash, uint64, uint64, error)
```

Comment:
This `GetTransaction` method is not great for several reasons. First, it requires us to have an index of txhash -> tx, which is strictly optional. The index we have is not complete.

Another problem is the amount of return values. A signature with this number of return values implies there may be more values returned in the future. If we wanted to change this, it would break the interface. So we need to either define a new struct which is to be returned by this method, or break it up into multiple methods.

I also wonder why a tracer would need to query transactions by hash. I think it's a niche use case and we might be better off not allowing it.

Reply:
This method will be used for the `trace_transaction` RPC, ref https://github.com/ethereum/go-ethereum/blob/c8dfa31422/eth/tracers/live/live_api_trace.go#L93-L116

Reply:
> I also wonder why a tracer would need to query transactions by hash.

I think a tracer will almost never need this in and out of itself. Only in the case where the tracer is exposing an API. Which makes me think if live tracers exposing an API is a good abstraction. It is easy to abuse the live tracing interface to expose an API that doesn't need tracing hooks at all. I wonder if we should introduce plugins that can also register live tracers and/or APIs instead.

Reply:
> I wonder if we should introduce plugins that can also register live tracers and/or APIs instead.

agree with it, plugins will be more flexible and useful for different use cases.